dlib/image_transforms/interpolation.h

```diff
@@ -965,6 +965,46 @@ namespace dlib
         swap(img, temp);
     }
 
+// ----------------------------------------------------------------------------------------
+
+    template <typename image_type>
+    point_transform_affine letterbox_image (
+        const image_type& img_in,
+        image_type& img_out,
+        long size
+    )
+    {
+        DLIB_CASSERT(size > 0, "size must be bigger than zero, but was " << size);
+        const_image_view<image_type> vimg_in(img_in);
+        image_view<image_type> vimg_out(img_out);
+
+        // the scaling factor of the image
+        const auto scale = size / std::max<double>(vimg_in.nr(), vimg_in.nc());
+
+        // early return if the image has already the requested size and no padding is needed
+        if (scale == 1 && vimg_in.nr() == vimg_in.nc())
+        {
+            assign_image(vimg_out, vimg_in);
+            return point_transform_affine();
+        }
+
+        // black background
+        vimg_out.set_size(size, size);
+        assign_all_pixels(vimg_out, 0);
+
+        // resize the image so that it fits into a size x size image
+        image_type temp;
+        image_view<image_type> vtemp(temp);
+        vtemp.set_size(std::round(scale * vimg_in.nr()), std::round(scale * vimg_in.nc()));
+        resize_image(vimg_in, temp);
```

Comment:
To make this even faster you could ask resize_image to write directly into the output image by using sub_image.  Then you won't have to do that extra image copy at the end.

Reply:
I will try that later! We might end up with a function that is nicer than I expected.
At the beginning I was hesitant to send this PR because the function is somewhat trivial, but I ended up learning a lot, as always. Thank you so much! :D

Reply:
No problem :)

Reply:
I've removed the extra copy, but I didn't notice any performance improvement. However, I feel better with myself.
I wonder if I also should do the same with the background assignment, just assign the padding, instead of the whole image...

Reply:
Cool.  Yeah, using zero_border_pixels() to zero just the border pixels might be faster.

Reply:
Oh, I never noticed the existence of `zero_border_pixels`, I knew about `assign_border_pixels`. The fact that `zero_border_pixels`takes a rectangle is very convenient!